File: front/src/__tests__/amrpRenderer.test.ts
import { describe, expect, it } from "vitest";
import { aimMessagePlainText, normalizeAIMessage } from "@/conversation/rendering/amrp";
import { renderMarkdownSegment, splitMarkdownSegments } from "@/conversation/rendering/markdown/markdownEngine";

describe("AMRP renderer", () => {
  it("normalizes legacy rich message fields", () => {
    const message = normalizeAIMessage({
      id: "m1",
      role: "assistant",
      content: "结果如下",
      msgType: "tool_call",
      toolName: "read_file",
      toolArguments: { path: "src/App.vue" },
      toolResult: { ok: true },
      status: "success",
      createdAt: "2026-09-20T00:00:00.000Z",
    });

    expect(message.markdown).toBe("");
    expect(message.blocks).toHaveLength(1);
    expect(message.blocks[0]).toMatchObject({
      kind: "tool",
      name: "read_file",
      arguments: { path: "src/App.vue" },
      result: { ok: true },
      status: "success",
    });
    expect(aimMessagePlainText(message)).toContain("read_file");
  });

  it("splits fences by renderer type", () => {
    const segments = splitMarkdownSegments([
      "正文",
      "```dart",
      "void main() {}",
      "```",
      "```diff",
      "- old",
      "+ new",
      "```",
      "```terminal",
      "$ flutter test",
      "```",
      "```mermaid",
      "graph TD",
      "A --> B",
      "```",
      "```html-preview",
      "<h1>Hello</h1>",
      "```",
    ].join("\n"));

    expect(segments.map((segment) => segment.type)).toEqual([
      "markdown",
      "code",
      "diff",
      "terminal",
      "mermaid",
      "html-preview",
    ]);
    expect(segments[1].language).toBe("dart");
  });

  it("normalizes persisted reasoning duration", () => {
    const message = normalizeAIMessage({
      id: "m2",
      role: "assistant",
      content: "回复",
      reasoningContent: "思考",
      reasoningDurationMs: 2850,
      createdAt: "2026-09-20T00:00:00.000Z",
    });

    expect(message.thinking?.duration).toBe(2.85);
  });

  it("没有 reasoning 内容时不伪造 Thinking Block", () => {
    const pending = normalizeAIMessage({
      id: "generating-request",
      role: "assistant",
      content: "",
      status: "streaming",
    });
    const completed = normalizeAIMessage({
      id: "assistant-1",
      role: "assistant",
      content: "回复",
      status: "sent",
    });

    expect(pending.thinking).toBeUndefined();
    expect(completed.thinking).toBeUndefined();
  });

  it("sanitizes unsafe markdown links and keeps supported structure", () => {
    const html = renderMarkdownSegment([
      "# 标题",
      "",
      "[危险](javascript:alert(1))",
      "",
      "| A | B |",
      "| - | - |",
      "| 1 | 2 |",
      "",
      "行内公式 $E=mc^2$",
    ].join("\n"));

    expect(html).toContain("<h1>");
    expect(html).toContain("amrp-table-scroll");
    expect(html).toContain("katex");
    expect(html).not.toContain('href="javascript:');
  });
  it("renders only registered citation ids as interactive references", () => {
    const withoutSources = renderMarkdownSegment("unknown [99]", []);
    expect(withoutSources).toContain("[99]");
    expect(withoutSources).not.toContain('data-citation-id="99"');

    const mixed = renderMarkdownSegment("known [1], unknown [99]", ["1"]);
    expect(mixed).toContain('data-citation-id="1"');
    expect(mixed).not.toContain('data-citation-id="99"');
  });

});

File: mobile_app/test/features/conversation/rendering/amrp_renderer_test.dart
import 'package:amitia_app/features/conversation/rendering/amrp.dart';
import 'package:amitia_app/features/conversation/rendering/markdown/markdown_parser.dart';
import 'package:amitia_app/shared/models/models.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('旧消息可以归一化为 AMRP Tool Block', () {
    final message = AmrpMessage.fromChatMessage(
      ChatMessage(
        id: 'm1',
        role: MessageRole.assistant,
        type: MessageType.toolCall,
        content: '',
        time: DateTime(2026, 9, 20),
        toolName: 'read_file',
        toolResult: '{"ok":true}',
      ),
    );

    expect(message.blocks, hasLength(1));
    final block = message.blocks.single as AmrpToolBlock;
    expect(block.name, 'read_file');
    expect(block.status, AmrpToolStatus.success);
    expect(block.result, '{"ok":true}');
    expect(message.plainText, contains('read_file'));
  });

  test('Markdown Fence 按 Renderer 类型拆分', () {
    final segments = splitAmitiaMarkdown(
      [
        '正文',
        '```dart',
        'void main() {}',
        '```',
        '```diff',
        '- old',
        '+ new',
        '```',
        '```terminal',
        r'$ flutter test',
        '```',
        '```mermaid',
        'graph TD',
        'A --> B',
        '```',
        '```html-preview',
        '<h1>Hello</h1>',
        '```',
        r'$$',
        'E=mc^2',
        r'$$',
      ].join('\n'),
    );

    expect(segments.map((segment) => segment.type).toList(), [
      AmitiaMarkdownSegmentType.markdown,
      AmitiaMarkdownSegmentType.code,
      AmitiaMarkdownSegmentType.diff,
      AmitiaMarkdownSegmentType.terminal,
      AmitiaMarkdownSegmentType.mermaid,
      AmitiaMarkdownSegmentType.htmlPreview,
      AmitiaMarkdownSegmentType.latex,
    ]);
    expect(segments[1].language, 'dart');
  });

  test('未闭合 Fence 保持流式状态', () {
    final segments = splitAmitiaMarkdown('```dart\nclass Test {');
    expect(segments.single.type, AmitiaMarkdownSegmentType.code);
    expect(segments.single.streaming, isTrue);
  });

  test('完整消息状态映射到 AMRP', () {
    AmrpMessage map(MessageStatus status) => AmrpMessage.fromChatMessage(
      ChatMessage(
        id: status.name,
        role: MessageRole.assistant,
        type: MessageType.text,
        content: 'ok',
        time: DateTime(2026, 9, 20),
        status: status,
      ),
    );

    expect(map(MessageStatus.queued).state, AmrpMessageState.queued);
    expect(map(MessageStatus.streaming).state, AmrpMessageState.streaming);
    expect(map(MessageStatus.delivered).state, AmrpMessageState.completed);
    expect(map(MessageStatus.interrupted).state, AmrpMessageState.interrupted);
    expect(map(MessageStatus.error).state, AmrpMessageState.failed);
    expect(map(MessageStatus.cancelled).state, AmrpMessageState.cancelled);
  });

  test('系统提示按系统消息渲染并保留内容', () {
    final message = AmrpMessage.fromChatMessage(
      ChatMessage(
        id: 'system-notice',
        role: MessageRole.assistant,
        type: MessageType.systemNotice,
        content: '模型响应失败',
        time: DateTime(2026, 9, 20),
        status: MessageStatus.error,
      ),
    );

    expect(message.role, AmrpMessageRole.system);
    expect(message.markdown, '模型响应失败');
  });
}
